### ISSUE TITLE
phone state sensor stopped working after android 14 update

### ISSUE THREAD

**Issue by ea5522** [home-assistant/android#4040]
**Home Assistant Android app version(s):**
2023.11.2-full
**Android version(s):**
14
**Device model(s):**
sm-g998u
**Home Assistant version:**
latest
**Last working Home Assistant release (if known):**
issue happend after update to android 14
**Description of problem, include YAML if issue is related to notifications:**
phone state sensor stopped working after update to android 14

**Companion App Logs:**
not needed

**Screenshot or video of problem:**

**Additional information:**

**Comment by ea5522** [comment]
Hi Daniel, I isolated the issue to gsm calls and some apps. Whatsapp and my voip app works but gsm stopped working. Telegram calls never worked. Where do I find companion app logs?

**Comment by dshokouhi** [comment]
> Where do I find companion app logs?

Settings > companion app > troubleshooting > show and share logs

make sure to open the app, reproduce the issue and then get the logs

**Comment by ea5522** [comment]
Ok, seems like it works when companion app is opened. When minimized, it works 50/50. Possibly an android issue. I did check the battery optimization settings, etc..

**Comment by dshokouhi** [comment]
Samsung devices are known to add multiple different battery optimization settings under different names like Power Saving, Data Saving etc... check all your device settings to see if a new one was added in Android 14

**Comment by ea5522** [comment]
Yeah, I checked them all already. Strange, i'll report back if i find something

**Comment by ccristal** [comment]
Exactly same issue here on a Galaxy S21 after Android14 update. Phone State sensor gets updated immediately as soon as the HA companion app is brought to foreground. If it's not, it's a hit and miss.

**Comment by dshokouhi** [comment]
> I have disabled all the optimizations and enabbled every permission I could find under the Settings->Apps->Home Assistant, to no avail.

have you checked things like power saving, data saving etc...? Those are Samsung specific settings and they will not be found under the HA app.

your device is indeed getting intents as expected so looks like the OS is not sending them when screen is off so most likely there is another new setting added that needs to be disabled.

`12-31 13:19:01.884  5320  5320 D SensorReceiver: Received intent: android.intent.action.PHONE_STATE`

**Comment by jpelgrom** [comment]
> your device is indeed getting intents as expected so looks like the OS is not sending them when screen is off 

Which [could be explained by an Android 14 behavior change] (https://developer.android.com/about/versions/14/behavior-changes-all#pending-broadcasts-queued), nothing you can do about that.

**Comment by dshokouhi** [comment]
> > your device is indeed getting intents as expected so looks like the OS is not sending them when screen is off
> 
> Which [could be explained by an Android 14 behavior change](https://developer.android.com/about/versions/14/behavior-changes-all#pending-broadcasts-queued), nothing you can do about that.

yup and that points again to battery optimizations or some kind of power saving setting. One should expect the app to remain in `foreground_service` state when in the background to avoid it being in the `cached` state. Here is the graph of the app importance sensor when the app has proper background access there shoudl be 2 states `foreground_service` and `foreground`

![image](https://github.com/home-assistant/android/assets/1634145/f58aac3c-144d-4684-9edd-ec20c87e6f18)

![image](https://github.com/home-assistant/android/assets/1634145/0bbf8f56-dfed-4ab3-adde-10ff2bf3a181)

**Comment by ccristal** [comment]
> > I have disabled all the optimizations and enabbled every permission I could find under the Settings->Apps->Home Assistant, to no avail.
> 
> have you checked things like power saving, data saving etc...? Those are Samsung specific settings and they will not be found under the HA app.
> 
> your device is indeed getting intents as expected so looks like the OS is not sending them when screen is off so most likely there is another new setting added that needs to be disabled.
> 
> `12-31 13:19:01.884 5320 5320 D SensorReceiver: Received intent: android.intent.action.PHONE_STATE`

Yes, the "offhook" event was notified immediately, as I explained in my post. The called ended at 13:19:10, but the "idle" event was notified only when the app was brought back to foreground (i.e. at 13:19:16). And we're not talking about the screen being off here, just the HA app not being in foreground.

Does the HA app do [this](https://developer.android.com/about/versions/14/changes/fgs-types-required)?

**Comment by dshokouhi** [comment]
> Does the HA app do [this](https://developer.android.com/about/versions/14/changes/fgs-types-required)?

no we do not target android 14 yet

Based on the description and the link that jpelgrom posted it looks like the OS is limiting the broadcasts sent to your device as `offhook` needs to be sent before `idle` and theres probably some type of cached queue its hitting. I think it may be worth it to check device settings more.

dontkillmyapp.com usually  has some good tips but not sure it was updated with most recent Samsung changes.

**Comment by ccristal** [comment]
> Based on the description and the link that jpelgrom posted it looks like the OS is limiting the broadcasts sent to your device as `offhook` needs to be sent before `idle` and theres probably some type of cached queue its hitting. I think it may be worth it to check device settings more.

The behaviour is completely random. Sometimes offhook is sent and idle is not, sometimes it's the other way round. Most often neither is sent. And yes, I have checked all the settings I could find, to no avail. Those are the following:

- Power saving
- Protect battery
- Adaptive battery
- Put unused apps to sleep

Curiously, I can't find the HA app in the list of those that can be added to "Never auto sleeping apps".

**Comment by TPMunster** [comment]
Just wanted to pitch in saying that I'm having the same issues with my Pixel 8 Pro ✌️

**Comment by dshokouhi** [comment]
> Just wanted to pitch in saying that I'm having the same issues with my Pixel 8 Pro ✌️

did you make sure to grant the app background access?

my pixel 8 pro is working as expected with battery optimizations disabled.

![image](https://github.com/home-assistant/android/assets/1634145/ec3ef887-5939-4421-9fea-86bff6f77c67)

offhook and ringing state both comes in.

![image](https://github.com/home-assistant/android/assets/1634145/a526927b-110e-454e-b6af-7bc8840e6af3)

**Comment by matthewgutz** [comment]
Running into the same issue on S22 (SM-S901U) _**android**_ 14.
Doing some research, could this be the issue?
`If your app targets Android 14, it must specify appropriate foreground service types.`
[](url)https://developer.android.com/about/versions/14/changes/fgs-types-required###

**Comment by jpelgrom** [comment]
> Doing some research, could this be the issue? `If your app targets Android 14, it must specify appropriate foreground service types.`

The app does not yet target Android 14, nor does it use a foreground service for this sensor, so that seems unlikely.

**Comment by albob75** [comment]
For info, I'm having similar issues with my Samsung Galaxy S23 Ultra, and my wife's Pixel 6a. Both were working great for a good few months, but a while back the Pixel became "un-reliable" updating it's sensors, and now my S23 ultra is doing the same. Both are now on Android 14, so possibly the issues started when they updated.

Seems I now have the same behaviour as @ea5522, if the HA android app is open the sensors update immediately, but as soon as I minimize it or switch to another app it's hit or miss as to whether the sensors update or not.

I've checked on both phones and battery access for HA is set to un-restricted, and no power savings modes are activated. I have also tried turning off adaptive battery, but nothing seems to work.

FYI - sometimes I can see that the app importance has gone into cached, which shouldn't happen with un-restricted battery, is that correct?

**Comment by albob75** [comment]
> > Based on the description and the link that jpelgrom posted it looks like the OS is limiting the broadcasts sent to your device as `offhook` needs to be sent before `idle` and theres probably some type of cached queue its hitting. I think it may be worth it to check device settings more.
> 
> The behaviour is completely random. Sometimes offhook is sent and idle is not, sometimes it's the other way round. Most often neither is sent. And yes, I have checked all the settings I could find, to no avail. Those are the following:
> 
> * Power saving
> * Protect battery
> * Adaptive battery
> * Put unused apps to sleep
> 
> Curiously, I can't find the HA app in the list of those that can be added to "Never auto sleeping apps".

@ccristal  - I think when you set the app battery access to "un-restricted" it removes the app from the "Never auto sleep apps" list, and doesn't let you select it. So possibly they are doing the same thing?

**Comment by hatchling** [comment]
I want to echo a basically identical experience to what @albob75 is reporting. In my case the phones involved are a Pixel 5 and Pixel 6, Android 14. I did similarly notice that app state is going to "cached" when not opened despite removing all restrictions I could find. 

My automation was very reliable with either phone previously, and tracked charging state and type (it turns off all my lights when either phone is placed on a Pixel charging stand).

**Comment by Ragingfire105** [comment]
Having similar sensor update issues on my Pixel 7.  Charging type is being used to trigger a scene and isn't updating unless I open the app.

**Comment by Oni159** [comment]
Same issue on my side, with my phone (Pixel 7) and wife's phone (Pixel 8); Android 14.
When I unplugged the phone, home assistant can take 10 minutes to understand the phone is unplugged.

**Comment by dshokouhi** [comment]
Although I am unable to reproduce the issue I have an idea for a potential fix but it requires end user testing. Please install the debug APK from the link below by downloading the artifact and extracting the zip. Run both production and debug side by side for a few days, the apps can run in parallel and the debug one will have a red icon to help distinguish it. Make sure to give the device a different name so you can remove the entry when the test is done. Make sure to grant the app background access but also keep the debug app usage minimal, use the blue icon to access your dashboards. We want to test how well it does when the app is in teh background untouched, like what has been described in this issue. Also please check the Phone State and Battery Charging sensors for this test. If possible in a few days show us a bar chart/graph like below to help show its updating better or the same. Keeping production and debug running at the same time will help us see if its behaving better or not.

https://github.com/home-assistant/android/actions/runs/8010690680

![image](https://github.com/home-assistant/android/assets/1634145/1d8fdc9d-0e25-46f0-b1c4-21aa763ff221)

**Comment by hatchling** [comment]
Unfortunately I'm still experiencing long delays with the "instant" sensors, like battery state. This first image is similar to what you posted earlier, and may or may not be directly helpful because of the long duration.
![image](https://github.com/home-assistant/android/assets/1924086/6411a7de-9c3a-4314-9555-14c97c9e0e83)

This second image shows a specific test. I plugged into AC at 6:21, and noticed it took 5 mins to reflect on the dev app, and 10 mins on the normal app. I suspect (without looking at code) that they are just on their own independent 15 minute update interval. The main thing is that the sensor wasn't reported instantly. Maybe worth noting that the app state seems to like going into `cached` state for 15 mins after plugging in, then it returns to `foreground_service`. I did not touch the phone at all during this time, and the screen was off when I plugged it in.
![app_testing](https://github.com/home-assistant/android/assets/1924086/3ea21260-f8df-4805-9668-702088a63bf9)

I'm happy to test any further builds if you find anything else, **thanks again for trying to fix this for us**.

**Comment by ea5522** [comment]
I switched to galaxy s24 ultra. Sensor is working fine on android 14.

**Comment by dshokouhi** [comment]
> If I clear the recent apps list

this will stop the sensor worker from retaining its schedule and is not recommended

**Comment by siw1973** [comment]
Most master users of HA use the Companion App daily, bringing it to the foreground therefore wouldn't experience a Android 14 "timeout".

For reference and context my wife and child have experienced "location lockout" that seems to occur after 6 days.

This impacts Alarm and automated door lock consequences.

**Comment by natebgamer** [comment]
Just wanted to add I'm having these same issues posted above on a Samsung s21 and Samsung a52s since they were updated to android 14.

My automations that were using the phones being on charge as a trigger have started being delayed/not running at all/work Instantly if I then bring the app to the foreground just like posted above.

Done all the things already mentioned like turning off all battery saving options and making sure the app has background access etc.

Hopefully this gets fixed soon.

**Comment by siw1973** [comment]
Is this something the developers need to do with the app classification in Play Store?

https://9to5google.com/2023/05/05/samsung-background-android-14/

**Comment by dshokouhi** [comment]
> Is this something the developers need to do with the app classification in Play Store?
> 
> https://9to5google.com/2023/05/05/samsung-background-android-14/

no that is not required for the app, the app also is not targeting android 14

please keep in mind its working for some users but not all.

**Comment by siw1973** [comment]
When you say the app is not targeting Android 14 what do you mean? Android 14 has been out 5 months. 

I have the same settings on two Pixel 7 phones with the same Home Assistant settings across the Apo and Android. The only difference is that my wife opens the app in the foreground it works. Then closes the app after which the sensor services run for a differing amount of time and then die.

I use the App frequently and then close the App. But will open it at least three / four times a day.

It's clearly an Android change that breaks Home Assistant on Android 14.

**Comment by dshokouhi** [comment]
> When you say the app is not targeting Android 14 what do you mean? Android 14 has been out 5 months.

Its developer terminology about what SDK version the app is targeting.  This is not anything an end user would know about unless they develop android applications.

I think the [comment made earlier](https://github.com/home-assistant/android/issues/4040#issuecomment-1879070161) still stands.

**Comment by siw1973** [comment]
OK, from a semi laymans terms you are saying that the Home Assistant Companion App does not support Android 14. If that's the case what is on the Product Roadmap for this.

I'm was assuming the Companion App pushes to HA when a Sensor Update is detected, be that WiFi Network, Battery Change. Which is rather different than your earlier test to see if the Companion App accepts OS messages.

**Comment by dshokouhi** [comment]
> OK, from a semi laymans terms you are saying that the Home Assistant Companion App does not support Android 14.

no we are not saying that at all. 

> I'm was assuming the Companion App pushes to HA when a Sensor Update is detected, be that WiFi Network, Battery Change or Location. Which is rather different than your earlier test to see if the Companion App accepts OS messages.

yes the OS is supposed to fire an intent which the app listens to that triggers the update, as you can see from the troubleshooting above that intent is not reaching these devices. thats what the linked comment is referring to.

**Comment by siw1973** [comment]
So I had an idea and so far so good.

I've added a simple entity state widget to the home screen which seems to stop Android from terminating Home Assistant as it will need to continously keep the HA connection open even if I have closed the HA Companion App.

Will continue to monitor and report back in a week.

**Comment by siw1973** [comment]
This failed yet again.

Honestly not sure what to do here or if anyone is actually going to pick this up.

Makes the Companion App pretty useless for Android initiated routines for people who don't regularly go into the app.

**Comment by dshokouhi** [comment]
> Makes the Companion App pretty useless for Android initiated routines for people who don't regularly go into the app.

FWIW my wife hardly ever opens the app and the sensor works fine for her and the intents are always received

just a thought but on most devices when the phone starts ringing the screen turns on, does enabling the interactive sensor make the situation better?

**Comment by siw1973** [comment]
Hi, I'll set it up and have a try. Thanks for the suggestion.

Is this an Android initiated trigger to the Home Assistant Companion App?  Therefore Android will insist on keeping the connection open.

The location sensor is set up to update every minute due to arrival routines but I think that's an HA Companion App initiated trigger that Android may not "care" about. 

That's the reason I thought that the Widget approach would have worked as it was an Android "owned" process that it wouldn't shut down the HA "connection"

**Comment by dshokouhi** [comment]
> Is this an Android initiated trigger to the Home Assistant Companion App? Therefore Android will insist on keeping the connection open.

the interactive sensor uses [native android intents](https://github.com/home-assistant/android/blob/master/app/src/main/java/io/homeassistant/companion/android/HomeAssistantApplication.kt#L99) much like the [phone state sensor](https://github.com/home-assistant/android/blob/master/app/src/main/java/io/homeassistant/companion/android/HomeAssistantApplication.kt#L139) to push updates



> The location sensor is set up to update every minute due to arrival routines but I think that's an HA Companion App initiated trigger that Android may not "care" about.

location tracking uses Gogles Fused Location provider that is native to android devices that are play store certified, the app [provides an intent](https://github.com/home-assistant/android/blob/master/app/src/full/java/io/homeassistant/companion/android/sensors/LocationSensorManager.kt#L958) that updates come to



> That's the reason I thought that the Widget approach would have worked as it was an Android "owned" process that it wouldn't shut down the HA "connection"

so widgets like template and entity state subscribe to a websocket connection when the screen turns on and closes the connection when the screen turns off in an effort to keep the widgets as up to date as possible. Its tough to say whats happening here honestly. 

One other thing you may try is looking at logcat from an app like logcat reader or android studio to see if there are additional logs at the time of occurrence something there may help shed a light to the issue.

**Comment by siw1973** [comment]
Trialing interactive sensor now 👍

**Comment by dshokouhi** [comment]
@siw1973  did the interactive sensor make the situation better?

**Comment by siw1973** [comment]
It seems to have as I've had no lost location tracking, but I've introduced an Octopus Energy Agile tab on our Energy Dashboard that my wife is more engaged with so I'm unsure as to whether it is the interactive sensor or the introduction of an HA Dashboard that she is actually checks.

Will report back later, but ir's a start.

**Comment by zgbee** [comment]
So I just tried to enable a sensor yesterday for the current WiFi network, and also noticed this problem that many other people seem to have where the sensor only updates if I bring the Home Assistant companion app to the foreground. I have a Pixel 7a running Android 14.

After seeing the discussion here, I enabled the app importance sensor and put a little test widget on a dashboard. I also turned on the interactive sensor (but am not monitoring anywhere). After the initial Unknown from when App Importance was first added to my dashboard but hadn't received an update yet - I have 2 statuses: `cached` (which lasted for 10 seconds), and then `perceptible` (where it's been for the past 10 minutes).

I've already double-checked the background usage/etc settings.. is it expected that I should only ever see `foreground` or `foreground_service` statuses?

**Comment by zgbee** [comment]
I think I spoke too soon.. I didn't verify with the logs at the time, but yesterday it seemed like my automation tied to the "instant" WiFi sensor was triggered 10-15 minutes after I rejoined the network.

**Comment by zgbee** [comment]
So in looking at the history for my phone's App Importance, it looks like just now when I arrived home it was in a state of `foreground`, and about 10 minutes later when I was off doing something else, the same second the state changed to `foreground_service`, the automation I have set to trigger on my phone's wifi sensor logged an execution.

I did not open the Home Assistant app, but it's possible the change to `foreground_service` occurred while unlocking my phone. It may have been in `foreground` before that as when I arrived home, I used a companion app widget on my Android phone's home screen to unlock the building's front door (which is out of my wifi range).

**Comment by dshokouhi** [comment]
@zgbee did you grant the app background access too?

**Comment by zgbee** [comment]
@dshokouhi as far as I know I've granted the proper background access. In my App battery usage settings, I have Allow background usage on, which has the description of "Enable for real-time updates, disable to save battery"

I went into the app preferences and took a screenshot of the All Permissions screen, in case that's helpful for verifying my settings: https://imgur.com/a/Lnf1Uqi

**Comment by siw1973** [comment]
> It seems to have as I've had no lost location tracking, but I've introduced an Octopus Energy Agile tab on our Energy Dashboard that my wife is more engaged with so I'm unsure as to whether it is the interactive sensor or the introduction of an HA Dashboard that she is actually checks.
> 
> Will report back later, but ir's a start.

I've actually had no drops after adding the "compelling" element to one of the HA Dashboards and also adding the Interaction Sensor. I can't really validate which is keeping the app from closing in the background, but I don't have the Interaction Sensor on my phone, but of course I access the HA CA frequently.

@zgbee what phone do you have, assuming not a Pixel ?

**Comment by zgbee** [comment]
@siw1973 by the "compelling" element to you need a card for the Interactive sensor? I had previously added one for the App Importance sensor, but not interaction. I'm not home for a couple days, but I've added an Interaction card to one of my dashboards and will try to test behavior this weekend.

I'm running Android 14 (currently build AP1A.240405.002) on a Pixel 7a

**Comment by siw1973** [comment]
Hi there. unfortunatley the "compelling" element was actually an Octopus Agile Energy Rate card my Energy Dashboard. This she is actually interested in and actually opens the HA Companion App maybe every couple of days when looking at the optimum time to use the household appliances. We both open this a lot now as it can save upto 90% energy dependent on when we run the washing machine / dishwasher / dryer etc..

I added the Interaction Sensor to both of our Phone "Status" cards as well to see how often that triggers.

The point I was making was that I can't tell which of the two changes has lead to the stable nature of the HA CA not being closed down by Android. Mine never closed down, but I always look at my HA CA at least 5 times a day so I suspect it's the actual opening of the App in the foreground that retains it running in the background.

With respect to the Pixel 7a, I have a Pixel 7 running the same build and don't see half of the options you show on your screen shot when accessing App Permissions ? Through which route do you access them ?

**Comment by zgbee** [comment]
Ah, ok. Well.. it'll be interesting to see if adding the interaction sensor to my dashboard has any effect.

As for the permissions, it was a hidden screen. I went into App info (via the press-hold icon and selecting App info from the menu), then Permissions, and then using the 3-dot menu icon in top-right to select All permissions (didn't realize that existed till yesterday, ha)

**Comment by siw1973** [comment]
Found it thank you 🤣 Learnt something new today! 

I have exactly the same permissions, good to know.

**Comment by natebgamer** [comment]
I added the interactive card to my dashboard. At first it was working every time even when my phone was locked but on. It would report "On" but like the issues above, after a while it seems all sensors just stop updating due to the app being put to sleep or whatever. I recently moved to a newer phone (S21>S22) so redid all permissions etc and still have the same issues with sensors stopping working after a while if the app hasn't been opened recently.

**Comment by dshokouhi** [comment]
> I added the interactive card to my dashboard. At first it was working every time even when my phone was locked but on. It would report "On" but like the issues above, after a while it seems all sensors just stop updating due to the app being put to sleep or whatever. I recently moved to a newer phone (S21>S22) so redid all permissions etc and still have the same issues with sensors stopping working after a while if the app hasn't been opened recently.

no the issue is not that all sensors stop updating, thats definitely a background access issue. The issue here is that some intents do not arrive in order to update the phone state sensor quickly. Your issue is unrelated here. Please file a new issue and make sure to go through all device settings to turn off any and all battery /power/data savings.

**Comment by natebgamer** [comment]
> > I added the interactive card to my dashboard. At first it was working every time even when my phone was locked but on. It would report "On" but like the issues above, after a while it seems all sensors just stop updating due to the app being put to sleep or whatever. I recently moved to a newer phone (S21>S22) so redid all permissions etc and still have the same issues with sensors stopping working after a while if the app hasn't been opened recently.
> 
> no the issue is not that all sensors stop updating, thats definitely a background access issue. The issue here is that some intents do not arrive in order to update the phone state sensor quickly. Your issue is unrelated here. Please file a new issue and make sure to go through all device settings to turn off any and all battery /power/data savings.

Sorry yes I misspoke with saying "Stop working" instead of delayed, but my issues are definitely the same as all those listed above where sometimes it can take a while for the sensors to update.

**Comment by dshokouhi** [comment]
> but my issues are definitely the same as all those listed above where sometimes it can take a while for the sensors to update.

That's really not the same here. One state that does not appear long is ringing. If the phone misses that intent then indeed the state will be missing. As far as I know the other states are working. Your issue described still sounds like background access if it impacts all sensors for a while.

**Comment by siw1973** [comment]
I had this go down on the wife's phone last week. Looks like what was shiny, new and compelling is no longer compelling...... 

I know this is annoying, but is it possible to force the app into the foreground once a week from Home Assistant to ensure it does not close ?

**Comment by dshokouhi** [comment]
> I know this is annoying, but is it possible to force the app into the foreground once a week from Home Assistant to ensure it does not close ? I know this is a total hack, but I'd like to have this reliably working.

https://companion.home-assistant.io/docs/notifications/notification-commands#webview

this command will open the app

**Comment by hatchling** [comment]
Just to add a data point:

After migrating from my affected Pixel 5 to a new Pixel 7 Pro (which is on Android 14) I no longer experience this issue. The Pixel 7 charger type state updates instantly.

My wife's Pixel 6 is still affected.

I suspect there is more to this issue than strictly the Android version but unfortunately don't have hard data to back this up.

**Comment by zgbee** [comment]
fwiw, adding a card for the interaction sensor to my dashboard didn't seem to have any effect. I'm travelling right now so it has to wait a few days, but I'm actually not sure if I've tried manually opening the home assistant app just before I expect the automation to trigger - i.e. if the app is in foreground and phone is unlocked, will the sensor update instantly?

**Comment by siw1973** [comment]
@dshokouhi Unfortunately I don't agree with the title change. I have the Companion App stopping sending any information to HA, not instantly, but full stop.

Of course if the App is brought to the foreground that is a separate "hack" and is something that never used to have to be done when I was on Android 13.

**Comment by dshokouhi** [comment]
> Unfortunately I don't agree with the title change. I have the Companion App stopping sending any information to HA, not instantly, but full stop.
> 
> Of course if the App is brought to the foreground that is a separate "hack" and is something that never used to have to be done when I was on Android 13.

Then that is not the same issue and you should create a new one with logs. This issue is about a device with missing intents that are used to update instantly. Full stop means there is something else at play. Every 15 minutes the sensor worker runs and that does not use intents and will pull the actual data unless there is an error. So full stop is not the same issue here. I think some users are conflating the actual issue with other similar issues here.

**Comment by rezadast2** [comment]
Adding my name to the list here where I have observed the exact same behavior as discuss in this thread. Pixel 8 Pro on Android 14. Background access is granted. Battery Optimization is disabled. Sensors are updated immediatley when the App is in the foreground. Once the app is in the background, App Importance goes to "cached" and my sensors are not updated immediately and can take up to 15 minutes. 

I noticed this behavior as soon as I got this Pixel 8 Pro. My previous phone was a Pixel 6 Pro on A13 with no issues. I believe there was a 2 week period where my P6Pro was on A14 and did not experience this behavior, however I cannot guarantee that.

I made a reddit thread in /r/homeassistant [here](https://www.reddit.com/r/homeassistant/comments/18fbyqs/pixel_8_pro_phone_sensors_degraded/), not thinking that github was the better place to report my findings. I just assumed the App needed to be updated so I left it alone for a few months. But as my automations get more complex, this issue is starting to be quite annoying, so here I am.

As i read through this thread, this comment below looks to have some promise, but I dont know what this really does:

> I have the same problem on my Nothing Phone (1) with Android 14. I checked that there was no battery optimization and that the HA app has background access. The HA app goes into cached mode and then the sensors don't update unless I'm launching the HA app.
> 
> I managed to solve the problem in part as follows:
> 
> * activate the hidden developer options
> * going to "suspend execution for cached apps"
> * select Disabled (was on "device default")
> * restart
> 
> Then the sensors update as expected ONLY if I leave HA in the recent app screen. If I clear the recent apps list, I have to launch HA again to update the sensors.

When I change the following setting in Developer Options: "Suspend execution for chached apps" > Disabled, My sensors are now updated instantly even when app is in background. I've only tested this for 15 minutes or so, but when app is in background, the App Importance is still "cached" but my sensors are updated instantly when I plug/unplug my phone to the charger.

I doubt this is the real solution to the problem, but wanted to call out my findings.

**Comment by zgbee** [comment]
I've kinda given up on this ever working. Even if I have the Home Assistant app open, with screen unlocked, and watch my phone join my home WiFi network via the menu bar icon change.. the automation that should trigger still doesn't seem to happen for another 5-10 minutes.

**Comment by tankdeer** [comment]
I stumbled upon this issue trying to determine if what I've been experiencing the past few weeks is known issue or not. 

I am experiencing the same issue as others. Pixel 7. Android 14. Build AP2A.240705.004

In my case it's the charging/charger type sensors that I've noticed, but it seems like it's all "updates instantly" sensors. As mentioned I've only really noticed this the past few weeks, maybe a month. I think like many I use charging status to trigger both good night & good morning routines, and this has made that very annoying. 

What's more disappointing is the apparent lack of interest from the devs to research further. I think we all realize this is an open source project, and that devs are always stretched thin, and I for one certainly appreciate that. I think it's also painfully obvious that this is _not_ a configuration issue in the sense that it's not a couple users who accidentally turned on battery saver or disabled background usage. This is dozens of people reporting the same thing in an area that has worked without issue for, in most cases, several years. There is obviously a compatibility issue with the Companion App & Android 14 - _something_ has changed. As for what I have no idea. Nor do I know why some people have issues and others don't, but this is only going to get worse unless a real solution is found.

I think like many I am happy to debug, provide logs, whatever. But we need somebody actively looking into this on the HA side. Keeping the app open all the time is not really a viable workaround.

**Comment by dshokouhi** [comment]
> What's more disappointing is the apparent lack of interest from the devs to research further.

I dont know where you got this from, we have been unable to reproduce the issue and have made attempts to fix the issue and stay active responding to users like yourself. Unless the developers can reproduce the issue teh next best thing is for a user who knows kotlin and is experiencing the issue can help contribute and provide a fix.

**Comment by tankdeer** [comment]
> > What's more disappointing is the apparent lack of interest from the devs to research further.
> 
> I dont know where you got this from, we have been unable to reproduce the issue and have made attempts to fix the issue and stay active responding to users like yourself. Unless the developers can reproduce the issue teh next best thing is for a user who knows kotlin and is experiencing the issue can help contribute and provide a fix.

Perhaps that came out wrong, and for that I apologize. It just seems like this is often being dismissed as a configuration/battery saver type issue, when I truly believe that is not the case. Numerous people have confirmed they everything configured correctly and still have the issue.

**Comment by dshokouhi** [comment]
> Perhaps that came out wrong, and for that I apologize. It just seems like this is often being dismissed as a configuration/battery saver type issue, when I truly believe that is not the case. Numerous people have confirmed they everything configured correctly and still have the issue.

I get that but at the same time numerous people have it working just fine, so yea on the surface it may appear to be a configuration issue. Often times when you are troubleshooting an issue thats pretty much where you start if you are unable to replicate the issue.

**Comment by tankdeer** [comment]
> Often times when you are troubleshooting an issue thats pretty much where you start if you are unable to replicate the issue.

Starting there sure, but unless there are some hidden settings that numerous people have overlooked, myself included, it seems like that is not the case. (Trust me, I would love nothing more than for this to be that simple)

Personally I have a suspicion this is an android bug or other internal functionality change that is only manifesting on certain devices. But of course that is just a guess and I would not know if it was something that could be coded around

**Comment by tankdeer** [comment]
I am back and ready to eat some crow. 😬 

I am about 95% sure I have found and resolved the issue. It was a setting, however one that was not mentioned earlier. In my case it seems like a lot of my server settings on the companion app had been wiped out including importantly here - "Persistent connection". It was set to "Never". The description implies it's only used to send notifications, but it seems like underneath it's likely used for more than that. In my case I set it to "On home WiFi only", since that's all I really care about, and now everything instantly updates again, even with the app closed, phone locked, idle for hours, etc.

I'd recommend anybody having this issue check your settings as well. They are located in the companion app:

Settings -> Companion App -> Servers & Devices (Click on your HA server) -> Persistent Connection.

In my case my home WiFI SSIDs and internal connection URLs had also been lost, not sure what happened.

Hope it helps.

**Comment by Ragingfire105** [comment]
I held off on reporting that mine has been working lately because I wasn't sure why.  It seems we had the same issue I just hadn't put it together until I read this.  I was working on a notifications issue when I discovered mine had been also reset to none for the persistent notification as well.

I do have to wonder, is this meant to be a requirement?  If not I hope this helps get to the bottom of the issue.

**Comment by hatchling** [comment]
> I do have to wonder, is this meant to be a requirement? If not I hope this helps get to the bottom of the issue.

A persistent connection is not required for this feature to work. My fairly vanilla Pixel 7 Pro is set to "Persistent connection: Never" and it works perfectly and instantly regardless of the phone state.

If changing this helps, it could be taken as a partial workaround for some, but it's not the root cause.

**Comment by rezadast2** [comment]
Going to chime in here and report back from my findings from about a month ago. 

First, I want to address the setting "persistent connection" being the solution. Yes, turning this on made all my sensors update immediately, but this also absolutely killed my phone battery. So I would disagree this is a solution to the problem. Read my reddit thread I linked in my comment up above [here](https://github.com/home-assistant/android/issues/4040#issuecomment-2168210761).

Second, there was a user that posted earlier about going into the devolpoer options and change the following settings.

> activate the hidden developer options
> going to "suspend execution for cached apps"
> select Disabled (was on "device default")
> restart

I have no idea what this setting actually does, but I have had these settings on for over a month now, and this has fixed my issue. All sensors update instantly and I almost never open the Comapanion App (meaning it persists after rebooting my phone and still not opening the app my sensors are updated instantly). I also don't believe this to be the "Solution", but maybe this helps the developers in this thread on identifying the real issue.

**Comment by tankdeer** [comment]
In my experience "suspend execution for cached apps" made absolutely no change. 🤷

**Comment by zgbee** [comment]
So after all the recent chatter in this thread, I decided to revisit this and enabled developer mode so I could toggle the "Suspend execution for cached apps" to Disabled, reboot my phone, and as @rezadast2 mentioned load up a dashboard on my laptop to monitor the WiFi status of my phone (a sensor I had enabled months ago but turned off the automation for since it was essentially non-functional).

Despite not having opened the HA app on my phone recently, and the required reboot after the dev setting change, the WiFi card on my dashboard updated (essentially) immediately when toggling my phone's WiFi off and on.

**Comment by dshokouhi** [comment]
Just wanted to check in with the users who found a potential workaround/fix how has it been going now thats been over a month? are the sensors updating as you would expect now?

**Comment by tankdeer** [comment]
Still working perfectly here with persistent connection enabled. Yes it does seem to increase battery usage, but that is a tradeoff I was willing to make to have this functional. I may experiment further with it disabled, but for the time being I am happy

**Comment by dshokouhi** [comment]
> Still working perfectly here with persistent connection enabled. Yes it does seem to increase battery usage, but that is a tradeoff I was willing to make to have this functional. I may experiment further with it disabled, but for the time being I am happy

have you tried enabling various sensors other than the ones mentioned? the last notification sensor might help keep things awake without requiring the persistent connection. it requires a special permission when enabled.

**Comment by rezadast2** [comment]
Setting "suspend execution for cached apps" to disabled under developer options is still working for me as a workaround.

But I also still believe this should not be the solution to the problem. I am using this as a workaround in hopes a real solution is provided.

**Comment by dshokouhi** [comment]
> I did not try the last notification sensor. I can, however it states that it increases battery usage and if I am doing that, I don't see the advantage over persistent connection which also increases battery usage.

by default this sensor will not send any updates to HA server without an allow list. The note is more for users who disable teh allow list because that can send updates quite frequently. The persistent connection pings the server every 30 seconds. I think your phone may get less notifications than that. I would say its worth a shot. FWIW I am personally using this sensor to update me on food delivery status :) its one of my favorite sensors in the app :) 

from the docs: https://companion.home-assistant.io/docs/core/sensors#last-notification

`You need to either create an allow list or enable the setting to "Disable Allow List Requirement". Keep in mind without an allow list this sensor has the potential to drain a lot of battery. We highly recommend creating an allow list over disabling this requirement.`


> But I also still believe this should not be the solution to the problem. I am using this as a workaround in hopes a real solution is provided.

This is not an easy issue to work around because the app is getting intents not delivered by the OS. Which is the only real way we should be updating these sensors. Have you noticed any battery life issues since enabling that dev option?

**Comment by tankdeer** [comment]
I will play with the last notification sensor as a possible alternative, thanks

**Comment by tankdeer** [comment]
Following up after a few days. Per the above recommendation I disabled persistent connection and enabled the last notification sensor. I picked a couple random apps to add to the list since I don't currently have a use-case for them.

So far this has been an improvement. My sensors are back to updating real time, and battery life is improved over the persistent connection.

I am happy using this as a long term solution/workaround. Thanks

**Comment by dshokouhi** [comment]
thanks for teh update!!

> and battery life is improved over the persistent connection.

curious if you notice a difference from before this issue started if enabling this sensor added to the drain but i understand lots of time has passed so may not be easy to tell :)

**Comment by tankdeer** [comment]
> thanks for teh update!!
> 
> > and battery life is improved over the persistent connection.
> 
> curious if you notice a difference from before this issue started if enabling this sensor added to the drain but i understand lots of time has passed so may not be easy to tell :)

Hard to say for sure since as you said it's been a while. I'd say it's likely comparable to what it was before this all started. Close enough at least.

**Comment by rezadast2** [comment]
> > Have you noticed any battery life issues since enabling that dev option?

I have not noticed any negative battery life since changing the dev option. 

On that note, wanted to check in with another data point. I have upgraded to a Pixel 9 Pro and experience the same behavior. Needed to change "suspend execution for cached apps" > disabled. After doing this, instant sensor updates when the app is not at the foreground.

**Comment by carlomoretto27** [comment]
Same issue with S22 and S22 ultra.

Only thing that works for me is persistent connection being on. Disabled the suspend execution for cached apps but that made no difference.

What other workarounds have people found as I don't want our batteries draining just to have location updates work

**Comment by dshokouhi** [comment]
> What other workarounds have people found as I don't want our batteries draining just to have location updates work

try to enable the last notification sensor and grant the permission, no need to set an allow list ot disable it unless you actually want to use the sensor. Granting the permission should be enough. A user reported earlier that worked for them and teh battery impact was negligible.

https://companion.home-assistant.io/docs/core/sensors#last-notification

**Comment by carlomoretto27** [comment]
> > What other workarounds have people found as I don't want our batteries draining just to have location updates work
> 
> try to enable the last notification sensor and grant the permission, no need to set an allow list ot disable it unless you actually want to use the sensor. Granting the permission should be enough. A user reported earlier that worked for them and teh battery impact was negligible.
> 
> https://companion.home-assistant.io/docs/core/sensors#last-notification

Thank you so much. Have just tried and it seems to be working. Hopefully it carries on working

**Comment by oystermon** [comment]
Hi just wanted to add my experience with what I think is a similar issue here.

I access my HA instance via my own VPN, and am using an Android device (Pixel 9)
I have a similar issue (sensors only update when app is open, or updates only when persistent connection is set to always)
The suggestion to turn on the last notification sensor seems to have solved my issue, and the "suspend execution for cached apps" also seemed to solve it too (I chose to do the first option as that seems a better fix)

Something that maybe of interest:
When attempting to diagnose my issue, I noticed my Next Alarm Sensor seems to work instantly, but not the WiFi Connection sensor or any of the Wifi related ones. When triggering the next alarm sensor (i.e. setting a different alarm time), all sensors gets updated at the same time. Not sure if that's a clue to what maybe happening? I assume it's to do with that additional notification permission that is requested when toggling the last notification sensor?

**Comment by tx71** [comment]
Since about 2 months, I have the exact same issue on a Pixel 9 Pro on Android 15. I notice it mainly with my Good Night automation, which is triggered by wireless charging state. I have to unlock my phone and open HA app for the automation to trigger. All battery saving options are disabled, HA has all possible permissions. 

I will try the 'last notification' workaround, and report back.

**Comment by siw1973** [comment]
This issue is still frustrating the life out of me and HA is still not connecting to provide accurate location to control arrival routines when leaving work and when arriving home if HA had not been opened recently.

My wife's battery ran out recently, and amazingly after never actually opening Lookout it fired an accurate location.

This seems to indicate that process triggers can stay alive if coded correctly.

**Comment by cridemichel** [comment]
Hi all,
I have a pixel 8 pro with android 16 up-to-date and I have same issue. I tried to enable "last notification sensor "but it seems it does not fix the issue. Now I disabled "suspend execution for cached apps" and see what happens. Is there any news on this very annoying issue?
best C.

**Comment by heeendrik25** [comment]
Confirming this issue is still present and observing behavior that strongly points to the app's background service being terminated by the OS.

**Environment**:
Device: Google Pixel 10 Pro XL
Android Version: 16 (QPR1)
Home Assistant Companion App Version: Latest stable from Play Store

**Behavior**:
The phone_state sensor update is severely delayed or fails to trigger entirely from the background.
When it does update from the background, the delay for the state to change from offhook to idle (after a call ends) can be up to 40 seconds or in most of the cases, the state does not update at all, in some cases it updates after ending a call. The update is only processed then when the Companion App is manually opened and brought to the foreground.

This makes automations dependent on this sensor unreliable.

**Troubleshooting Performed**:
The issue persists despite all recommended settings to enable background execution:
Battery usage for the app is set to "Unrestricted".
All relevant permissions (Phone) are granted.
The Android system setting "Remove permissions if app is unused" is disabled.

**Hypothesis**:
The fact that the update often requires opening the app is a strong indicator of a Pixel-specific background process limitation that is aggressively terminating the app's service, even when "Unrestricted" battery settings are selected.

Happy to provide logs or assist with further testing to help diagnose this.

**Comment by cridemichel** [comment]
I confirm the behavior which you observed, indeed disabling "suspend execution for cached apps" in developer options
seems to fix the issue in my case,

best C.

**Comment by heeendrik25** [comment]
Thanks @cridemichel for the input. I can also confirm that disabling "Suspend execution for cached apps" in Developer Options completely resolves the delay for me. After disabling it, the phone_state sensor updates are now instantaneous.

However, I agree this can only be a temporary workaround. This is a global, system-wide setting that affects every application on the device. While the Home Assistant app is well-optimized, I'm concerned about the potential battery drain from other, less well-behaved apps that will now also be allowed to execute while cached.

This raises a question for the developers: Is there any way for the Companion App to signal a higher priority for its background services to the Android OS, specifically for critical sensor updates like the phone state?

For example, could the app use a high-priority foreground service (perhaps with a persistent notification only visible during an active call) to prevent the OS from suspending it?

It would be ideal if the app could ensure its own reliability without requiring users to disable a major power-saving feature for their entire system. Please let us know if there's any potential for a more integrated solution in a future update. Thanks for all your work on this great app!

**Comment by cridemichel** [comment]
I have to update my situation, even with ""Suspend execution for cached apps" disabled, location updates can be delayed and automations based on them fail...it is really annoying

**Comment by omegasupreme3** [comment]
> > What other workarounds have people found as I don't want our batteries draining just to have location updates work
> 
> try to enable the last notification sensor and grant the permission, no need to set an allow list ot disable it unless you actually want to use the sensor. Granting the permission should be enough. A user reported earlier that worked for them and teh battery impact was negligible.
> 
> https://companion.home-assistant.io/docs/core/sensors#last-notification

I own a Google Pixel 8a with Android 17 and can confirm this actually works!

_Previously I tried setting Persistent Connection to "Always" which also works with near real-time updates, however batter life is noticeably impacted. It also hits the foreground service timeout wall (roughly 6 hours budget every 24 hours), meaning that the foreground service gets destroyed after 6 hours and no updates are sent from there on. So definitely not an option._
